modules/PhenoDigm.py

```diff
@@ -382,23 +418,49 @@ def generate_phenodigm_evidence_strings(self, score_cutoff):
-        """Dump the Spark evidence dataframe as a compressed JSON file. The order of the evidence strings is not
-        maintained, and they are returned in random order as collected by Spark."""
-        with tempfile.TemporaryDirectory() as tmp_dir_name:
-            (
-                self.evidence.coalesce(1).write.format('json').mode('overwrite')
-                .option('compression', 'org.apache.hadoop.io.compress.GzipCodec').save(tmp_dir_name)
-            )
-            json_chunks = [f for f in os.listdir(tmp_dir_name) if f.endswith('.json.gz')]
-            assert len(json_chunks) == 1, f'Expected one JSON file, but found {len(json_chunks)}.'
-            os.rename(os.path.join(tmp_dir_name, json_chunks[0]), evidence_strings_filename)
+    def generate_mouse_phenotypes_dataset(self):
+        """Generate the related mousePhenotypes dataset for the corresponding widget in the target object."""
+        mouse_phenotypes = (
+            # Extract base model-target associations.
+            self.disease_model_summary
+            .select('model_id', 'biologicalModelAllelicComposition', 'biologicalModelGeneticBackground',
+                    'targetInModelMgiId')
+            .distinct()
 
-    def generate_mouse_phenotypes_dataset(self, mouse_phenotypes_filename):
-        """Based on the evidence string, generate the related mousePhenotypes dataset for the corresponding widget in
-        the target object."""
+            # Add gene mapping information.
+            .join(self.gene_mapping, on='targetInModelMgiId', how='inner')
 
+            # Add mouse phenotypes.
+            .join(self.model_mouse_phenotypes, on='model_id', how='inner')
+            .join(self.mp_terms, on='mp_id', how='inner')
 
+            # Add literature references.
+            .join(self.literature, on=['model_id', 'targetInModelMgiId'], how='left')
+
+            # Rename fields.
+            .withColumnRenamed('mp_id', 'modelPhenotypeId')
+            .withColumnRenamed('mp_term', 'modelPhenotypeLabel')
+
+            # Join phenotype class information.
+            .join(self.mp_class, on='modelPhenotypeId', how='inner')
+        )
+        # Post-process model ID field.
+        self.mouse_phenotypes = self._cleanup_model_identifier(mouse_phenotypes)
+
+    def write_datasets(self, evidence_strings_filename, mouse_phenotypes_filename):
+        """Dump the Spark evidence dataframe as a compressed JSON file. The order of the evidence strings is not
+        maintained, and they are returned in random order as collected by Spark."""
+        for dataset, outfile in ((self.evidence, evidence_strings_filename),
+                                 (self.mouse_phenotypes, mouse_phenotypes_filename)):
+            logging.info(f'Processing dataset {outfile}')
+            with tempfile.TemporaryDirectory() as tmp_dir_name:
+                (
+                    dataset.coalesce(1).write.format('json').mode('overwrite')
```

Comment:
I think considering the file size coalescing is a risky step. Even the score threshold is set to 80, the number of phenotypes is 550k+. When using the default cutoff (0) even a very memory rich VM failed.

Reply:
Hmm, just out of curiosity, how much memory were you using? It's just that it seems to work on my desktop (64 GB RAM) even without any threshold. Although I haven't measured how much memory is actually used.

Also, I think we need to revisit this question in general (can discuss during the stand-up). This needs to be uniform across all parsers: they need to output either a single JSON file (this could be done without coalescing by just concatenating all JSON chunks together), or to always output a directory with JSONs.